File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_cpp/code_output/BookManagementDB.cpp
#include <sqlite3.h>
#include <stdexcept>
#include <string>
#include <vector>

class SQLException : public std::runtime_error {
public:
    explicit SQLException(const std::string& message) : std::runtime_error(message) {}
};

class BookManagementDB {
private:
    struct DatabaseHandle {
        sqlite3* db;
        DatabaseHandle(const std::string& dbName) {
            int rc = sqlite3_open(dbName.c_str(), &db);
            if (rc != SQLITE_OK) {
                throw SQLException("Cannot open database: " + std::string(sqlite3_errmsg(db)));
            }
        }
        ~DatabaseHandle() {
            if (db) sqlite3_close(db);
        }
        operator sqlite3*() { return db; }
    };

    struct Statement {
        sqlite3_stmt* stmt;
        DatabaseHandle& db;
        Statement(sqlite3_stmt* stmt, DatabaseHandle& db) : stmt(stmt), db(db) {}
        ~Statement() {
            if (stmt) sqlite3_finalize(stmt);
        }
        operator sqlite3_stmt*() { return stmt; }
    };

    DatabaseHandle db;

    void createTable() {
        const char* sql = "CREATE TABLE IF NOT EXISTS books ("
                         "id INTEGER PRIMARY KEY AUTOINCREMENT, "
                         "title TEXT, "
                         "author TEXT, "
                         "available INTEGER)";
        char* errMsg = nullptr;
        if (sqlite3_exec(db, sql, nullptr, nullptr, &errMsg) != SQLITE_OK) {
            throw SQLException(errMsg ? std::string(errMsg) : "Failed to create table");
            sqlite3_free(errMsg);
        }
    }

    Statement prepare(const std::string& sql) {
        Statement stmt(sqlite3_prepare_v2(db, sql.c_str(), -1, nullptr, nullptr), db);
        if (stmt.stmt == nullptr) {
            throw SQLException(sqlite3_errmsg(db));
        }
        return stmt;
    }

public:
    BookManagementDB(const std::string& dbName) : db(dbName) {
        createTable();
    }

    ~BookManagementDB() = default;

    void addBook(const std::string& title, const std::string& author) {
        auto stmt = prepare("INSERT INTO books (title, author, available) VALUES (?, ?, 1)");
        sqlite3_bind_text(stmt, 1, title.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 2, author.c_str(), -1, SQLITE_STATIC);
        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw SQLException(sqlite3_errmsg(db));
        }
    }

    void removeBook(int bookId) {
        auto stmt = prepare("DELETE FROM books WHERE id = ?");
        sqlite3_bind_int(stmt, 1, bookId);
        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw SQLException(sqlite3_errmsg(db));
        }
    }

    void borrowBook(int bookId) {
        auto stmt = prepare("UPDATE books SET available = 0 WHERE id = ?");
        sqlite3_bind_int(stmt, 1, bookId);
        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw SQLException(sqlite3_errmsg(db));
        }
    }

    void returnBook(int bookId) {
        auto stmt = prepare("UPDATE books SET available = 1 WHERE id = ?");
        sqlite3_bind_int(stmt, 1, bookId);
        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw SQLException(sqlite3_errmsg(db));
        }
    }

    struct Book {
        int id;
        std::string title;
        std::string author;
        int available;

        Book(int id, const std::string& title, const std::string& author, int available)
            : id(id), title(title), author(author), available(available) {}

        friend std::ostream& operator<<(std::ostream& os, const Book& book) {
            return os << "Book{ id=" << book.id
                      << ", title='" << book.title << "', "
                      << "author='" << book.author << "', "
                      << "available=" << book.available << " }";
        }
    };

    std::vector<Book> searchBooks() {
        std::vector<Book> books;
        auto stmt = prepare("SELECT * FROM books");
        if (sqlite3_step(stmt) != SQLITE_ROW) return books;

        int colCount = sqlite3_column_count(stmt);
        while (SQLITE_ROW == sqlite3_step(stmt)) {
            int id = sqlite3_column_int(stmt, 0);
            const char* title = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            const char* author = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
            int available = sqlite3_column_int(stmt, 3);
            books.emplace_back(id, title, author, available);
        }
        return books;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/BookManagementDB.py
import sqlite3
from typing import List

class BookManagementDB:
    def __init__(self, db_name: str):
        self.connection = sqlite3.connect(db_name)
        self.create_table()

    def create_table(self):
        create_table_sql = """
        CREATE TABLE IF NOT EXISTS books (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT,
            author TEXT,
            available INTEGER
        )
        """
        with self.connection:
            self.connection.execute(create_table_sql)

    def add_book(self, title: str, author: str) -> None:
        insert_sql = "INSERT INTO books (title, author, available) VALUES (?, ?, 1)"
        with self.connection:
            self.connection.execute(insert_sql, (title, author))

    def remove_book(self, book_id: int) -> None:
        delete_sql = "DELETE FROM books WHERE id = ?"
        with self.connection:
            self.connection.execute(delete_sql, (book_id,))

    def borrow_book(self, book_id: int) -> None:
        update_sql = "UPDATE books SET available = 0 WHERE id = ?"
        with self.connection:
            self.connection.execute(update_sql, (book_id,))

    def return_book(self, book_id: int) -> None:
        update_sql = "UPDATE books SET available = 1 WHERE id = ?"
        with self.connection:
            self.connection.execute(update_sql, (book_id,))

    def search_books(self) -> List['Book']:
        select_sql = "SELECT * FROM books"
        books = []
        with self.connection:
            cursor = self.connection.execute(select_sql)
            for row in cursor.fetchall():
                books.append(Book(*row))
        return books

    class Book:
        __slots__ = ('id', 'title', 'author', 'available')

        def __init__(self, id: int, title: str, author: str, available: int):
            self.id = id
            self.title = title
            self.author = author
            self.available = available

        def __repr__(self):
            return f"Book(id={self.id}, title='{self.title}', author='{self.author}', available={self.available})"

# Example usage
if __name__ == "__main__":
    db = BookManagementDB("library.db")
    db.add_book("Python Programming", "Eric Matthes")
    print(db.search_books())